lib/_stream_writable.js

```diff
@@ -489,83 +473,69 @@ function afterWrite(stream, state, count, cb) {
+
+  const { buffered, bufferedIndex, objectMode } = state;
+  const bufferedLength = buffered.length - bufferedIndex;
+
+  if (!bufferedLength) {
+    return;
+  }
+
+  let i = bufferedIndex;
+
   state.bufferProcessing = true;
-  let entry = state.bufferedRequest;
-
-  if (stream._writev && entry && entry.next) {
-    // Fast case, write everything using _writev()
-    const l = state.bufferedRequestCount;
-    const buffer = new Array(l);
-    const holder = state.corkedRequestsFree;
-    holder.entry = entry;
-
-    let count = 0;
-    let allBuffers = true;
-    while (entry) {
-      buffer[count] = entry;
-      if (entry.encoding !== 'buffer')
-        allBuffers = false;
-      entry = entry.next;
-      count += 1;
-    }
-    buffer.allBuffers = allBuffers;
+  if (bufferedLength > 1 && stream._writev) {
+    state.pendingcb -= bufferedLength - 1;
+
+    const callback = state.allNoop ? nop : (err) => {
+      for (let n = i; n < buffered.length; ++n) {
+        buffered[n].callback(err);
+      }
+    };
+    // Make a copy of `buffered` if it's going to be used by `callback` above,
+    // since `doWrite` will mutate the array.
+    const chunks = state.allNoop && i === 0 ? buffered : buffered.slice(i);
+    chunks.allBuffers = state.allBuffers;
 
-    doWrite(stream, state, true, state.length, buffer, '', holder.finish);
+    doWrite(stream, state, true, state.length, chunks, '', callback);
 
-    // doWrite is almost always async, defer these to save a bit of time
-    // as the hot path ends with doWrite
-    state.pendingcb++;
-    state.lastBufferedRequest = null;
-    if (holder.next) {
-      state.corkedRequestsFree = holder.next;
-      holder.next = null;
-    } else {
-      const corkReq = { next: null, entry: null, finish: undefined };
-      corkReq.finish = onCorkedFinish.bind(undefined, corkReq, state);
-      state.corkedRequestsFree = corkReq;
-    }
-    state.bufferedRequestCount = 0;
+    resetBuffer(state);
   } else {
-    // Slow case, write chunks one-by-one
-    while (entry) {
-      const chunk = entry.chunk;
-      const encoding = entry.encoding;
-      const cb = entry.callback;
-      const len = state.objectMode ? 1 : chunk.length;
-
-      doWrite(stream, state, false, len, chunk, encoding, cb);
-      entry = entry.next;
-      state.bufferedRequestCount--;
-      // If we didn't call the onwrite immediately, then
-      // it means that we need to wait until it does.
-      // also, that means that the chunk and cb are currently
-      // being processed, so move the buffer counter past them.
-      if (state.writing) {
-        break;
-      }
+    do {
+      const { chunk, encoding, callback } = buffered[i];
+      buffered[i++] = null;
+      const len = objectMode ? 1 : chunk.length;
+      doWrite(stream, state, false, len, chunk, encoding, callback);
+    } while (i < buffered.length && !state.writing);
+
+    if (i === buffered.length) {
+      resetBuffer(state);
+    } else if (i > 256) {
```

Comment:
Not this branch should rarely occur. Only if the chunks sent to the stream are very small relative to HWM.

Reply:
are you sure? wouldn't this happen every 256 writes unless the socket is fully flushed?

Reply:
If the source is faster than the `Writable` it will hit HWM and `'drain'` won't be emitted until it's fully flushed.

If the source is slower than the `Writable` then it will probably be able to be fully flush at some point.

If the source and the `Writable` are roughly the same then yes, it would occcur more commonly.

Reply:
here is a case that triggers it every 256 writes

```
const stream = require('stream')

const ws = new stream.Writable({
  write (data, enc, cb) {
    ws.write(Buffer.alloc(65536)) // just mimicking that a write is always pending like most RL cases (without using the cb)
    setTimeout(function () {
      cb(null)
    }, 1)
  }
})

ws.write(Buffer.alloc(65536))
ws.write(Buffer.alloc(65536))


```

Reply:
Um, that has some weird recursion there. Sorry, I can't wrap my head around that one.

Reply:
There are cases where it will trigger every 256 writes, yes.

Reply:
I think all real world cases will, as the writable stream is usually the slow one in a pipe. not that it's a problem though, it's only every 256 write.